mcs: get global tso from tso service primary addr failed

## Bug Report

<!-- Please answer these questions before submitting your issue. Thanks! -->

### What did you do?

run in api mode

<!-- If possible, provide a recipe for reproducing the error. -->

### What did you expect to see?

tikv/tidb can get tso, but cannot update gc safe point.

### What did you see instead?

```
[2023/04/18 11:51:29.565 +00:00] [WARN] [grpc_service.go:1494] ["trying to update gc safe point"] [old-safe-point=440217670660915200] [new-safe-point=0]
[2023/04/18 11:51:29.564 +00:00] [ERROR] [grpc_service.go:1999] ["get global tso from tso server failed"] [error="rpc error: code = Unavailable desc = transport is closing"]
```

### What version of PD are you using (`pd-server -V`)?

--- a/server/grpc_service.go
+++ b/server/grpc_service.go
@@ -20,6 +20,7 @@ import (
 	"io"
 	"path"
 	"strconv"
+	"strings"
 	"sync/atomic"
 	"time"
 
@@ -1787,7 +1788,12 @@ func (s *GrpcServer) getGlobalTSOFromTSOServer(ctx context.Context) (pdpb.Timest
 	})
 	ts, err := forwardStream.Recv()
 	if err != nil {
-		log.Error("get global tso from tso server failed", zap.Error(err))
+		log.Error("get global tso from tso service primary addr failed", zap.Error(err), zap.String("tso", forwardedHost))
+		if strings.Contains(err.Error(), "transport is closing") {
+			s.tsoClientPool.Lock()
+			delete(s.tsoClientPool.clients, forwardedHost)
+			s.tsoClientPool.Unlock()
+		}
 		return pdpb.Timestamp{}, err
 	}
 	return *ts.GetTimestamp(), nil

--- a/server/grpc_service.go
+++ b/server/grpc_service.go
@@ -1789,7 +1789,7 @@ func (s *GrpcServer) getGlobalTSOFromTSOServer(ctx context.Context) (pdpb.Timest
 	ts, err := forwardStream.Recv()
 	if err != nil {
 		log.Error("get global tso from tso service primary addr failed", zap.Error(err), zap.String("tso", forwardedHost))
-		if strings.Contains(err.Error(), "transport is closing") {
+		if strings.Contains(err.Error(), codes.Unavailable.String()) {
 			s.tsoClientPool.Lock()
 			delete(s.tsoClientPool.clients, forwardedHost)
 			s.tsoClientPool.Unlock()

--- a/server/grpc_service.go
+++ b/server/grpc_service.go
@@ -49,7 +49,8 @@ import (
 )
 
 const (
-	heartbeatSendTimeout = 5 * time.Second
+	heartbeatSendTimeout                   = 5 * time.Second
+	maxRetryTimesGetGlobalTSOFromTSOServer = 3
 )
 
 // gRPC errors
@@ -1774,29 +1775,40 @@ func (s *GrpcServer) getGlobalTSOFromTSOServer(ctx context.Context) (pdpb.Timest
 	if !ok || forwardedHost == "" {
 		return pdpb.Timestamp{}, ErrNotFoundTSOAddr
 	}
-	forwardStream, err := s.getTSOForwardStream(forwardedHost)
-	if err != nil {
-		return pdpb.Timestamp{}, err
-	}
-	forwardStream.Send(&tsopb.TsoRequest{
+	request := &tsopb.TsoRequest{
 		Header: &tsopb.RequestHeader{
 			ClusterId:       s.clusterID,
 			KeyspaceId:      utils.DefaultKeyspaceID,
 			KeyspaceGroupId: utils.DefaultKeyspaceGroupID,
 		},
 		Count: 1,
-	})
-	ts, err := forwardStream.Recv()
-	if err != nil {
-		log.Error("get global tso from tso service primary addr failed", zap.Error(err), zap.String("tso", forwardedHost))
-		if strings.Contains(err.Error(), codes.Unavailable.String()) {
-			s.tsoClientPool.Lock()
-			delete(s.tsoClientPool.clients, forwardedHost)
-			s.tsoClientPool.Unlock()
-		}
-		return pdpb.Timestamp{}, err
 	}
-	return *ts.GetTimestamp(), nil
+	var (
+		forwardStream tsopb.TSO_TsoClient
+		ts            *tsopb.TsoResponse
+		err           error
+	)
+	for i := 0; i < maxRetryTimesGetGlobalTSOFromTSOServer; i++ {
+		forwardStream, err = s.getTSOForwardStream(forwardedHost)
+		if err != nil {
+			return pdpb.Timestamp{}, err
+		}
+		forwardStream.Send(request)
+		ts, err = forwardStream.Recv()
+		if err != nil {
+			if strings.Contains(err.Error(), codes.Unavailable.String()) {
+				s.tsoClientPool.Lock()
+				delete(s.tsoClientPool.clients, forwardedHost)
+				s.tsoClientPool.Unlock()
+				continue
+			}
+			log.Error("get global tso from tso service primary addr failed", zap.Error(err), zap.String("tso-addr", forwardedHost))
+			return pdpb.Timestamp{}, err
+		}
+		return *ts.GetTimestamp(), nil
+	}
+	log.Error("get global tso from tso service primary addr failed after retry", zap.Error(err), zap.String("tso-addr", forwardedHost))
+	return pdpb.Timestamp{}, err
 }
 
 func (s *GrpcServer) getTSOForwardStream(forwardedHost string) (tsopb.TSO_TsoClient, error) {

